aws-api/src/main/java/com/amplifyframework/api/aws/AWSApiPlugin.java

```diff
@@ -736,4 +776,78 @@ private void transitionTo(ApiEndpointStatus newStatus) {
             }
         }
     }
+
+    /**
+     * Builds an {@link AWSApiPlugin}.
+     */
+    public static final class Builder {
+        private ApiAuthProviders apiAuthProviders;
+        private Map<String, List<Interceptor>> apiInterceptors;
+        private List<Interceptor> globalInterceptors;
+
+        private Builder() {
+            Builder.this.apiAuthProviders = ApiAuthProviders.noProviderOverrides();
+            Builder.this.globalInterceptors = new ArrayList<>();
+            Builder.this.apiInterceptors = new TreeMap<>();
+        }
+
+        /**
+         * Specify authentication providers.
+         * @param apiAuthProviders A set of authentication providers to use for API calls
+         * @return Current builder instance, for fluent construction of plugin
+         */
+        @NonNull
+        public Builder apiAuthProviders(@NonNull ApiAuthProviders apiAuthProviders) {
+            Objects.requireNonNull(apiAuthProviders);
+            Builder.this.apiAuthProviders = apiAuthProviders;
+            return Builder.this;
+        }
+
+        /**
+         * Add a global interceptor. The interceptor will be invoked for
+         * all API endpoints you've configured.
+         * Interceptors are applied in order. If you have multiple interceptors,
+         * the first-added interceptors get used first. Global interceptors
+         * are always applied first.
+         * @param interceptor An interceptor to apply to all API calls.
+         * @return A builder, to continue configuration of an AWSApiPlugin
+         */
+        @NonNull
+        public Builder addGlobalInterceptor(@NonNull Interceptor interceptor) {
```

Comment:
Hmm, this is exposing our OkHttp dependency in the public interface, which goes against our tenets, because it will prevent us from ever swapping out `OkHttp` for another HTTP client.  On the other hand, this is just in the plugin, not the category, and I think it's _unlikely_ that we'd ever swap it out (famous last words).

Would it be possible to create a facade `Interceptor` that decouples the public API from `OkHttp`.  Not sure if that's worth the effort or not.

Reply:
Hey @richardmcclellan! I deliberated on this same thing.

iOS does include a first party interceptor type in their *core* spec, @lawmicha and I chatted and concluded it didn't need to be a core spec.

IMO plugins themselves don't need much abstraction. The primary Inversion of Control in Amplify is along the Category/Plug-in boundary.

I can try a version with a first party interface, too, though.

Reply:
Update: I wrote an internal design doc and implemented a different solution. New solution is pushed. It addressed this feedback.